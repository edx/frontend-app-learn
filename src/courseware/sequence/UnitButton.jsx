import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Button } from '@edx/paragon';

import UnitIcon from './UnitIcon';
import CompleteIcon from './CompleteIcon';
import BookmarkFilledIcon from './bookmark/BookmarkFilledIcon';

export default function UnitButton({
  clickHandler,
  pageTitle,
  type,
  isActive,
  isBookmarked,
  isComplete,
  index,
}) {
  const onClick = useCallback(() => {
    clickHandler(index);
  });

  return (
    <Button
      className={classNames({
        active: isActive,
        'btn-outline-primary': !isActive,
        'btn-outline-secondary': isActive,
      })}

      onClick={onClick}
      title={pageTitle}
    >
      <UnitIcon type={type} />
      {isComplete ? <CompleteIcon className="text-success ml-2" /> : null}
      {isBookmarked ? (
        <BookmarkFilledIcon
          className="text-primary small position-absolute"
          style={{ top: '-3px', right: '5px' }}
        />
      ) : null}
    </Button>
  );
}

UnitButton.propTypes = {
  index: PropTypes.number.isRequired,
  isActive: PropTypes.bool,
  isBookmarked: PropTypes.bool,
  isComplete: PropTypes.bool,
  clickHandler: PropTypes.func.isRequired,
  pageTitle: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
};

UnitButton.defaultProps = {
  isActive: false,
  isBookmarked: false,
  isComplete: false,
};
