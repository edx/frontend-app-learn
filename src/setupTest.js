import 'core-js/stable';
import 'regenerator-runtime/runtime';
import '@testing-library/jest-dom';
import './courseware/data/__factories__';
import './course-home/data/__factories__';
import { getConfig, mergeConfig } from '@edx/frontend-platform';
import { configure as configureI18n } from '@edx/frontend-platform/i18n';
import { configure as configureLogging } from '@edx/frontend-platform/logging';
import { configure as configureAuth, getAuthenticatedHttpClient, MockAuthService } from '@edx/frontend-platform/auth';
import React from 'react';
import PropTypes from 'prop-types';
import { render as rtlRender } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { IntlProvider } from 'react-intl';
import MockAdapter from 'axios-mock-adapter';
import AppProvider from '@edx/frontend-platform/react/AppProvider';
import { reducer as courseHomeReducer } from './course-home/data';
import { reducer as coursewareReducer } from './courseware/data/slice';
import { reducer as modelsReducer } from './generic/model-store';
import { UserMessagesProvider } from './generic/user-messages';

import appMessages from './i18n';
import { fetchCourse, fetchSequence } from './courseware/data';
import executeThunk from './utils';
import buildSimpleCourseAndSequenceMetadata from './courseware/data/__factories__/sequenceMetadata.factory';

class MockLoggingService {
  logInfo = jest.fn();

  logError = jest.fn();
}

window.getComputedStyle = jest.fn(() => ({
  getPropertyValue: jest.fn(),
}));

export default function initializeMockApp() {
  mergeConfig({
    INSIGHTS_BASE_URL: process.env.INSIGHTS_BASE_URL || null,
    STUDIO_BASE_URL: process.env.STUDIO_BASE_URL || null,
    authenticatedUser: {
      userId: 'abc123',
      username: 'Mock User',
      roles: [],
      administrator: false,
    },
  });

  const loggingService = configureLogging(MockLoggingService, {
    config: getConfig(),
  });
  const authService = configureAuth(MockAuthService, {
    config: getConfig(),
    loggingService,
  });

  // i18n doesn't have a service class to return.
  configureI18n({
    config: getConfig(),
    loggingService,
    messages: [appMessages],
  });

  return { loggingService, authService };
}

window.scrollTo = jest.fn();

// MessageEvent used for indicating that a unit has been loaded.
export const messageEvent = {
  type: 'plugin.resize',
  payload: {
    height: 300,
  },
};

// Send MessageEvent indicating that a unit has been loaded.
export function loadUnit(message = messageEvent) {
  window.postMessage(message, '*');
}

let globalStore;

export async function initializeTestStore(options = {}, overrideStore = true) {
  const store = configureStore({
    reducer: {
      models: modelsReducer,
      courseware: coursewareReducer,
      courseHome: courseHomeReducer,
    },
  });
  if (overrideStore) {
    globalStore = store;
  }
  initializeMockApp();
  const axiosMock = new MockAdapter(getAuthenticatedHttpClient());
  axiosMock.reset();

  const {
    courseBlocks, sequenceBlock, courseMetadata, sequenceMetadata,
  } = buildSimpleCourseAndSequenceMetadata(options);

  const forbiddenCourseUrl = `${getConfig().LMS_BASE_URL}/api/courseware/course/${courseMetadata.id}`;
  const courseBlocksUrlRegExp = new RegExp(`${getConfig().LMS_BASE_URL}/api/courses/v2/blocks/*`);

  axiosMock.onGet(forbiddenCourseUrl).reply(200, courseMetadata);
  axiosMock.onGet(courseBlocksUrlRegExp).reply(200, courseBlocks);
  sequenceMetadata.forEach(metadata => {
    const sequenceMetadataUrl = `${getConfig().LMS_BASE_URL}/api/courseware/sequence/${metadata.item_id}`;
    axiosMock.onGet(sequenceMetadataUrl).reply(200, metadata);
  });

  axiosMock.onAny().reply((config) => {
    // eslint-disable-next-line no-console
    console.log(config.url);
    return [200, {}];
  });

  // eslint-disable-next-line no-unused-expressions
  !options.excludeFetchCourse && await executeThunk(fetchCourse(courseMetadata.id), store.dispatch);

  if (!options.excludeFetchSequence) {
    await Promise.all(sequenceBlock
      .map(block => executeThunk(fetchSequence(block.id), store.dispatch)));
  }

  return store;
}

function render(
  ui,
  {
    store = null,
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    return (
      // eslint-disable-next-line react/jsx-filename-extension
      <IntlProvider locale="en">
        <AppProvider store={store || globalStore}>
          <UserMessagesProvider>
            {children}
          </UserMessagesProvider>
        </AppProvider>
      </IntlProvider>
    );
  }

  Wrapper.propTypes = {
    children: PropTypes.node.isRequired,
  };

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// Re-export everything.
export * from '@testing-library/react';

// Override `render` method.
export {
  render,
};
