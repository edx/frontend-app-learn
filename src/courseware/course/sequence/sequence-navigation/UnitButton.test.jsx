import React from 'react';
import { render, screen, fireEvent } from '../../../../setupTest';
import UnitButton from './UnitButton';

describe('Unit Button', () => {
  const initialState = {
    models: {
      units: {
        other: {
          contentType: 'other',
          title: 'other-unit',
        },
        problem: {
          contentType: 'problem',
          title: 'problem-unit',
          complete: true,
          bookmarked: true,
        },
      },
    },
  };

  const mockData = {
    unitId: 'other',
    onClick: () => {},
  };

  it('hides title by default', () => {
    render(<UnitButton {...mockData} />, { initialState });
    expect(screen.getByRole('button')).not.toHaveTextContent('other-unit');
  });

  it('shows title', () => {
    render(<UnitButton {...mockData} showTitle />, { initialState });
    expect(screen.getByRole('button')).toHaveTextContent('other-unit');
  });

  it('does not show completion for non-completed unit', () => {
    const { container } = render(<UnitButton {...mockData} />, { initialState });
    container.querySelectorAll('svg').forEach(icon => {
      expect(icon).not.toHaveClass('fa-check');
    });
  });

  it('shows completion for completed unit', () => {
    const { container } = render(<UnitButton {...mockData} unitId="problem" />, { initialState });
    const buttonIcons = container.querySelectorAll('svg');
    expect(buttonIcons).toHaveLength(3);
    expect(buttonIcons[1]).toHaveClass('fa-check');
  });

  it('hides completion', () => {
    const { container } = render(<UnitButton {...mockData} unitId="problem" showCompletion={false} />, { initialState });
    container.querySelectorAll('svg').forEach(icon => {
      expect(icon).not.toHaveClass('fa-check');
    });
  });

  it('does not show bookmark', () => {
    const { container } = render(<UnitButton {...mockData} />, { initialState });
    container.querySelectorAll('svg').forEach(icon => {
      expect(icon).not.toHaveClass('fa-bookmark');
    });
  });

  it('shows bookmark', () => {
    const { container } = render(<UnitButton {...mockData} unitId="problem" />, { initialState });
    const buttonIcons = container.querySelectorAll('svg');
    expect(buttonIcons).toHaveLength(3);
    expect(buttonIcons[2]).toHaveClass('fa-bookmark');
  });

  it('handles the click', () => {
    const onClick = jest.fn();
    render(<UnitButton {...mockData} onClick={onClick} />, { initialState });
    fireEvent.click(screen.getByRole('button'));
    expect(onClick).toHaveBeenCalledTimes(1);
  });
});
