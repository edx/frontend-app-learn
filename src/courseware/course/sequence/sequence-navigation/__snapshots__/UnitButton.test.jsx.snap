// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit Button does not show bookmark 1`] = `
<DocumentFragment>
  <button
    class="btn"
    title="other-unit"
    type="button"
  >
    <img
      alt="fa-book"
      class="fa-book"
      data-testid="icon"
    />
  </button>
</DocumentFragment>
`;

exports[`Unit Button does not show completion for non-completed unit 1`] = `
<DocumentFragment>
  <button
    class="btn"
    title="other-unit"
    type="button"
  >
    <img
      alt="fa-book"
      class="fa-book"
      data-testid="icon"
    />
  </button>
</DocumentFragment>
`;

exports[`Unit Button hides completion 1`] = `
<DocumentFragment>
  <button
    class="btn"
    title="problem-unit"
    type="button"
  >
    <img
      alt="fa-edit"
      class="fa-edit"
      data-testid="icon"
    />
    <img
      alt="fa-bookmark"
      class="fa-bookmark"
      data-testid="icon"
    />
  </button>
</DocumentFragment>
`;

exports[`Unit Button hides title by default 1`] = `
<DocumentFragment>
  <button
    class="btn"
    title="other-unit"
    type="button"
  >
    <img
      alt="fa-book"
      class="fa-book"
      data-testid="icon"
    />
  </button>
</DocumentFragment>
`;

exports[`Unit Button shows bookmark 1`] = `
<DocumentFragment>
  <button
    class="btn complete"
    title="problem-unit"
    type="button"
  >
    <img
      alt="fa-edit"
      class="fa-edit"
      data-testid="icon"
    />
    <img
      alt="fa-check"
      class="fa-check"
      data-testid="icon"
    />
    <img
      alt="fa-bookmark"
      class="fa-bookmark"
      data-testid="icon"
    />
  </button>
</DocumentFragment>
`;

exports[`Unit Button shows completion for completed unit 1`] = `
<DocumentFragment>
  <button
    class="btn complete"
    title="problem-unit"
    type="button"
  >
    <img
      alt="fa-edit"
      class="fa-edit"
      data-testid="icon"
    />
    <img
      alt="fa-check"
      class="fa-check"
      data-testid="icon"
    />
    <img
      alt="fa-bookmark"
      class="fa-bookmark"
      data-testid="icon"
    />
  </button>
</DocumentFragment>
`;

exports[`Unit Button shows title 1`] = `
<DocumentFragment>
  <button
    class="btn"
    title="other-unit"
    type="button"
  >
    <img
      alt="fa-book"
      class="fa-book"
      data-testid="icon"
    />
    <span
      class="unit-title"
    >
      other-unit
    </span>
  </button>
</DocumentFragment>
`;
